panther_hardware_interfaces/src/roboteq_driver.cpp

```diff
@@ -0,0 +1,349 @@
+  }
+}
+
+void RoboteqDriver::SendRoboteqCmdChannel2(const std::int32_t cmd)
+{
+  try {
+    SyncSdoWrite<std::int32_t>(0x2000, 2, cmd);
+  } catch (const std::runtime_error & e) {
+    throw std::runtime_error(
+      "Error when trying to send channel 2 Roboteq command: " + std::string(e.what()));
+  }
+}
+
+void RoboteqDriver::ResetRoboteqScript()
+{
+  try {
+    SyncSdoWrite<std::uint8_t>(0x2018, 0, 2);
+  } catch (const std::runtime_error & e) {
+    throw std::runtime_error("Error when trying to reset Roboteq script: " + std::string(e.what()));
+  }
+}
+
+void RoboteqDriver::TurnOnEstop()
+{
+  // Cmd_ESTOP
+  try {
+    SyncSdoWrite<std::uint8_t>(0x200C, 0, 1);
+  } catch (const std::runtime_error & e) {
+    throw std::runtime_error("Error when trying to turn on estop: " + std::string(e.what()));
+  }
+}
+
+void RoboteqDriver::TurnOffEstop()
+{
+  // Cmd_MGO
+  try {
+    SyncSdoWrite<std::uint8_t>(0x200D, 0, 1);
+  } catch (const std::runtime_error & e) {
+    throw std::runtime_error("Error when trying to turn off estop: " + std::string(e.what()));
+  }
+}
+
+void RoboteqDriver::TurnOnSafetyStopChannel1()
+{
+  // Cmd_SFT Safety Stop
+  try {
+    SyncSdoWrite<std::uint8_t>(0x202C, 0, 1);
+  } catch (const std::runtime_error & e) {
+    throw std::runtime_error(
+      "Error when trying to turn on safety stop on channel 1: " + std::string(e.what()));
+  }
+}
+
+void RoboteqDriver::TurnOnSafetyStopChannel2()
+{
+  // Cmd_SFT Safety Stop
+  try {
+    SyncSdoWrite<std::uint8_t>(0x202C, 0, 2);
+  } catch (const std::runtime_error & e) {
+    throw std::runtime_error(
+      "Error when trying to turn on safety stop on channel 2: " + std::string(e.what()));
+  }
+}
+
+template <typename T>
+T RoboteqDriver::SyncSdoRead(const std::uint16_t index, const std::uint8_t subindex)
+{
+  std::unique_lock<std::mutex> sdo_read_lck(sdo_read_mtx_, std::defer_lock);
+  if (!sdo_read_lck.try_lock()) {
+    throw std::runtime_error(
+      "Can't submit new SDO read operation - the previous one is still being processed");
+  }
+
+  std::mutex mtx;
+  std::condition_variable cv;
+  T data;
+  std::error_code err_code;
+
+  // todo: In some cases (especially with frequencies higher than 100Hz, mostly during activation)
+  // deadlock can happen, when submitted function won't be executed and sdo_read_timed_out_ won't be
+  // set to false in result. Solution currently on hold - switching to PDO will also solve this
+  // issue
+  if (sdo_read_timed_out_) {
+    throw std::runtime_error(
+      "Can't submit new SDO read operation - previous one that timed out is still in queue");
+  }
+
+  try {
+    SubmitRead<T>(
+      index, subindex,
+      [&sdo_read_timed_out_ = sdo_read_timed_out_, &mtx, &cv, &err_code, &data](
+        std::uint8_t, std::uint16_t, std::uint8_t, std::error_code ec, T value) mutable {
+        // In this case function has already finished, and other variables don't exist
+        // and we have to end
```

Comment:
As for the comment, there was a possibility of weird behavior when I tried to make this operations RT safe - `cv.wait_for` could timeout, and while theoretically so should this operation on lely queue, it was still sometimes executed,  leading to references to destroyed variables. That said, as we will switch to PDO, and RT safe SDO are no longer needed, it will be removed in the `ros2-control-pdo-commands` PR. Lambda still stays, as this is callback passed to lely queue that executes upon finishing SDO operations, and lets now the main function that SDO operation finished, in result making it synchronous.